src/commands/auth.rs

```diff
@@ -5,22 +5,150 @@ use api::ApiClient;
+    use api::MockApiClient;
+    use credentials::MockCredentialsStorage;
+    use error::{ApiError, StorageError};
+    use models::User;
+    use tokio_test::assert_err;
+    use tokio_test::assert_ok;
+
+    const MOCK_API_TOKEN: &str = "SOME_API_TOKEN_VALUE";
+    const MOCK_EMAIL: &str = "toggl@user.org";
+    const MOCK_USER_NAME: &str = "Toggl User";
+
+    fn create_working_api_client() -> MockApiClient {
+        let mut api_client = MockApiClient::new();
+        let user = User {
+            api_token: MOCK_API_TOKEN.to_string(),
+            email: MOCK_EMAIL.to_string(),
+            fullname: Some(MOCK_USER_NAME.to_string()),
+            timezone: "UTC".to_string(),
+            default_workspace_id: 1,
+        };
+
+        api_client
+            .expect_get_user()
+            .returning(move || Ok(user.clone()));
+
+        return api_client;
+    }
+
+    fn create_failing_api_client() -> MockApiClient {
+        let mut api_client = MockApiClient::new();
+        api_client
+            .expect_get_user()
+            .returning(|| Err(Box::new(ApiError::Network)));
+
+        return api_client;
+    }
+
+    fn create_working_credentials_storage() -> MockCredentialsStorage {
+        let mut credentials_storage = MockCredentialsStorage::new();
+        credentials_storage.expect_persist().returning(|_| Ok(()));
+        return credentials_storage;
+    }
+
+    fn create_failing_credentials_storage() -> MockCredentialsStorage {
+        let mut credentials_storage = MockCredentialsStorage::new();
+        credentials_storage
+            .expect_persist()
+            .returning(|_| Err(Box::new(StorageError::Write)));
+        return credentials_storage;
+    }
+
+    #[tokio::test]
+    async fn a_valid_api_call_and_a_working_storage_returns_an_ok_result() {
+        // Arrange
+        let mut output = Vec::new();
+        let api_client = create_working_api_client();
+        let credentials_storage = create_working_credentials_storage();
+
+        // Act
+        let result =
+            AuthenticationCommand::execute(&mut output, api_client, credentials_storage).await;
+
+        // Assert
+        assert_ok!(result);
+    }
+
+    #[tokio::test]
+    async fn a_valid_api_call_and_a_working_storage_prints_a_success_message() {
+        // Arrange
+        let mut output = Vec::new();
+        let api_client = create_working_api_client();
+        let credentials_storage = create_working_credentials_storage();
+
+        // Act
+        let _ = AuthenticationCommand::execute(&mut output, api_client, credentials_storage).await;
+
+        // Assert
+        let expected_output = format!(
+            "{} {}",
+            AUTH_SUCCEEDED_MESSAGE.green(),
+            MOCK_EMAIL.green().bold()
+        );
+        let actual_output = String::from_utf8(output)
+            .expect(format!("empty output when {} was expected", expected_output).as_str());
+
+        assert_eq!(expected_output, actual_output);
```

Comment:
```suggestion
        let actual_output =  String::from_utf8(output).unwrap();

        assert_eq!(expected_output, actual_output, "empty output when {} was expected", expected_output);
```

An alternative - `assert_eq` will format additional arguments as the failure message, I don't know what will be better to work with in the long run though 🤔 

In fact, I'm not sure if the current code is actually printing the message when it fails?

Reply:
It'd only print when the unwrapping failed, In your case the unwrapping would fail without a message, so I think the former is better 🤔